akka-stream/src/main/scala/akka/stream/Attributes.scala

```diff
@@ -35,11 +35,29 @@ import akka.util.LineNumbers
  *
  * Operators should in general not access the `attributeList` but instead use `get` to get the expected
  * value of an attribute.
+ *
+ * Constructor is internal Akka API, use factories in companion to create instances.
  */
-final case class Attributes(attributeList: List[Attributes.Attribute] = Nil) {
+final class Attributes private[akka] (
+    val attributeList: List[Attributes.Attribute],
+    private val mandatoryAttributes: Map[Class[_], Attributes.MandatoryAttribute])
+    extends scala.Product
+    with scala.Serializable
+    with scala.Equals {
 
   import Attributes._
 
+  // for binary compatibility
+  @deprecated("Use factories on companion object instead", since = "2.8.0")
+  def this(attributeList: List[Attributes.Attribute] = Nil) =
+    this(attributeList, attributeList.foldLeft(Map.empty[Class[_], Attributes.MandatoryAttribute]) {
+      case (acc, attribute) =>
+        attribute match {
+          case m: Attributes.MandatoryAttribute if !acc.contains(m.getClass) => acc.updated(m.getClass, m)
```

Comment:
is this the most efficient way to build the Map? Maybe it is because we have to use the first attribute in the list and then a MapBuilder can't be used?

Reply:
Yeah, but I have thought about maybe just reverse building it instead with a builder, that way the right value "wins" and lists should anyway mostly be short in this constructor. 

It is not super important though, because for the default set of attributes we now invoke this just once when the materializer is created, rather than on each stream materialization like earlier, and then the `and`-methods are used instead in actual attribute changes on the stream.

Reply:
Rewrote it with a reversed iteration instead anyway, please take a second look.

Reply:
Lol, already forgot I asked or a second look and merged. I'll do a follow up PR if you spot anything.